xmipptomo/protocols/protocol_extract_subtomos.py

```diff
@@ -0,0 +1,222 @@
+        tsId = ts.getTsId()
+
+        # Defining the output folder
+        tomoPath = self._getExtraPath(tsId)
+        os.mkdir(tomoPath)
+
+        tomoFn = inTomograms[tomId].getFileName()
+
+        fnCoords = self.writeMdCoordinates(inTomograms[tomId], tomoPath)
+
+        params = ' --tomogram %s' % tomoFn
+        params += ' --coordinates %s' % fnCoords
+        params += ' --boxsize %i' % self.boxSize.get()
+        if self.invertContrast.get():
+            params += ' --invertContrast'
+        params += ' --subtomo'
+        params += ' --threads %i' % 1
+
+        params += ' -o %s' % tomoPath
+        self.runJob('xmipp_tomo_extract_subtomograms', params)
+
+        self.tomoFiles.append(tomoFn)
+
+    def createOutputStep(self):
+
+        outputSet = None
+        self.outputSubTomogramsSet = self._createSetOfSubTomograms(self._getOutputSuffix(SetOfSubTomograms))
+        self.outputSubTomogramsSet.setSamplingRate(self.tomograms.get().getSamplingRate())
+        self.outputSubTomogramsSet.setCoordinates3D(self.coords)
+        if self.tomograms.get().getFirstItem().getAcquisition():
+            acquisition = TomoAcquisition()
+            acquisition.setAngleMin(self.tomograms.get().getFirstItem().getAcquisition().getAngleMin())
+            acquisition.setAngleMax(self.tomograms.get().getFirstItem().getAcquisition().getAngleMax())
+            acquisition.setStep(self.tomograms.get().getFirstItem().getAcquisition().getStep())
+            self.outputSubTomogramsSet.setAcquisition(acquisition)
+
+        counter = 0
+
+        for item in self.tomograms.get().iterItems():
+            for ind, tomoFile in enumerate(self.tomoFiles):
+                if os.path.basename(tomoFile) == os.path.basename(item.getFileName()):
+                    coordSet = self.lines[ind]
+                    tomId = item.getObjId()
+                    tsId = self.tomograms.get()[tomId].getTsId()
+                    outputSet, counter = self.readSetOfSubTomograms(tomoFile,
+                                                                    self.outputSubTomogramsSet,
+                                                                    coordSet, 1, counter, tsId)
+
+        self._defineOutputs(outputSetOfSubtomogram=outputSet)
+        self._defineSourceRelation(self.coords, outputSet)
+
+    def readSetOfSubTomograms(self, tomoFile, outputSubTomogramsSet, coordSet, factor, counter, tsId):
+
+        self.info("Registering subtomograms for %s" % tomoFile)
+
+        outRegex = os.path.join(self._getExtraPath(tsId), pwutils.removeBaseExt(tomoFile)+'-*.mrc')
+
+        subtomoFileList = sorted(glob.glob(outRegex))
+
+        for idx, subtomoFile in enumerate(subtomoFileList):
+
+            self.debug("Registering subtomogram %s - %s" % (counter, subtomoFile))
+
+            subtomogram = SubTomogram()
+            subtomogram.cleanObjId()
+            subtomogram.setLocation(subtomoFile)
+            subtomogram.setCoordinate3D(coordSet[idx])
```

Comment:
You are asumming set of coordinates is not a sub set).

This will fail in this case... you can't use idx (I think) to identify the coordinates.

Reply:
well, is not failing because of the subset, but your subtomoFileList is sorted by name so:

1
10
11
2
3
...

whereas coordinates do not match this same order.